### DISCUSSION TITLE
2.4.0

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#3003] [General]
## What's Changed

* Change settings tab elements to be more compact
* Add clip skip slider
* Add select for custom VAE
* Add new style "Random Style"
* Update default anime model to animaPencilXL_v310
* Add button to reconnect the UI after Fooocus crashed without having to configure everything again (no page reload required)
* Add performance "hyper-sd" (based on [Hyper-SDXL 4 step LoRA](https://huggingface.co/ByteDance/Hyper-SD/blob/main/Hyper-SDXL-4steps-lora.safetensors))
* Add [AlignYourSteps](https://research.nvidia.com/labs/toronto-ai/AlignYourSteps/) scheduler by Nvidia, see 
* Add [TCD](https://github.com/jabir-zheng/TCD) sampler and scheduler (based on sgm_uniform)
* Add NSFW image censoring (disables intermediate image preview while generating). Set config value `default_black_out_nsfw` to True to always enable.
* Add argument `--enable-describe-uov-image` to automatically describe uploaded images for upscaling
* Add inline lora prompt references with subfolder support, example prompt: `colorful bird <lora:toucan:1.2>`
* Add size and aspect ratio recommendation on image describe
* Add inpaint brush color picker, helpful when image and mask brush have the same color
* Add automated Docker image build using Github Actions on each release, see [available images](https://ghcr.io/lllyasviel/fooocus)
* Add full raw prompts to history logs
* Change code ownership from @lllyasviel to @mashb1t for automated issue / MR notification

<details><summary>Details</summary>
<p>

* feat: support download huggingface files from a mirror site by @chenxinlong in https://github.com/lllyasviel/Fooocus/pull/2637
* chore: update interposer from v3.1 to v4.0 by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2717
* feat: add button to reconnect UI without having to reload the page by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2727
* feat: add optional model VAE select by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2867
* feat: choose random style by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2855
* feat: update anime from animaPencilXL_v100 to animaPencilXL_v310 by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2454
* refactor: rename label for reconnect button by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2893
* feat: add full raw prompt to history log by @docppp in https://github.com/lllyasviel/Fooocus/pull/1920
* fix: use correct border radius css property by @khanvilkarvishvesh in https://github.com/lllyasviel/Fooocus/pull/2845
* fix: do not close meta tag in HTML header by @e52fa787 in https://github.com/lllyasviel/Fooocus/pull/2740
* feat: automatically describe image on uov image upload by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1938
* add nsfw image censoring via config and checkbox by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/958
* feat: add align your steps scheduler by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2905
* feat: add support for lora inline prompt references by @cantor-set in https://github.com/lllyasviel/Fooocus/pull/2323
* feat: add tcd sampler and discrete distilled tcd scheduler based on sgm_uniform (same as lcm) by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2907
* feat: add performance hyper-sd based on 4step LoRA by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2812
* fix: remove leftover code for hyper-sd testing by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2959
* feat: optimize model management of nsfw image censoring by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2960
* feat: progress bar improvements by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2962
* feat: inline lora optimisations by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2967
* feat: change code owner from @lllyasviel to @mashb1t by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2948
* feat: only use valid inline loras, add subfolder support by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2968
* feature: Reads the size and ratio of the image and gives the recommended size by @xhoxye in https://github.com/lllyasviel/Fooocus/pull/2971
* feat: build and push container image for ghcr.io, update docker.md, and other related fixes  by @xynydev in https://github.com/lllyasviel/Fooocus/pull/2805
* feat: adjust line ending default config by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2991
* feat: add translation for image size describe by @mashb1t and @xhoxye in https://github.com/lllyasviel/Fooocus/pull/2992
* feat: read value 'CFG Mimicking from TSNR' from presets by @Alexdnk in https://github.com/lllyasviel/Fooocus/pull/2990
* feat: add inpaint brush color picker by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2997
* feat: remove labels from most of the image input fields by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2998
* feat: add clip skip handling by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2999
* feat: make ui settings more compact by @Alexdnk and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2590

</p>
</details> 

**Full Changelog**: https://github.com/lllyasviel/Fooocus/compare/2.3.1...v2.4.0

**Comment by ILikeToasters** [comment]
Video of almost all changes (someone snuck in the interface change last second...lol). Hopefully this will help explain some of the details for people if I didn't get anything wrong: https://youtu.be/8uqcdCF2Esk

**Reply by mashb1t** [reply]
@ILikeToasters yeah, let's call it "agile development" ^^

**Reply by MeAmCat** [reply]
idk what you guys did, but my image quality somehow doubled. Nice work. I am confused about the Hyper-SD thingy tho. Is it enabled by default or where do I enable it?

Edit: nvm I'm stupid found it.

**Reply by DavidDragonsage** [reply]
@ILikeToasters  So you "are" [Kleebz Tech AI](https://www.youtube.com/@KLEEBZTECH). Your channel is absolutely the best video source for Fooocus information! Thank you so much! ✨👍

**Reply by ILikeToasters** [reply]
That would be me. You are welcome and thanks for watching.

**Reply by BassAzayda** [reply]
> @ILikeToasters So you "are" [Kleebz Tech AI](https://www.youtube.com/@KLEEBZTECH). Your channel is absolutely the best video source for Fooocus information! Thank you so much! ✨👍

Without a doubt!

**Reply by AnalyticsXXI** [reply]
> That would be me. You are welcome and thanks for watching.

Oh, then I am your fun too xD. Just started to watch your guides about Fooocus, very good videos, you really understand how to explain things.

**Reply by stevewoh2** [reply]
i am new here, do we have to download every update every to get this to work?

**Reply by mashb1t** [reply]
@stevewoh2 no, just follow the installation instructions in the readme for your OS.

**Comment by cassiebug23** [comment]
For some reason my photos are coming out quite different with this new update. My custom checkpoints aren't as high quality as they were with the previous version/no longer fully resemble the people they were trained on. I was making pictures before it updated and things looked fine, so I'm not sure what's happened. All my settings are the same as they were pre-update.

**Reply by ILikeToasters** [reply]
Could you test some other images that used other models maybe? Reason I asked, after a couple comments on my Youtube video about the update I tested an image (JuggV8) and could not recreate. But then I just tested one from a different model (CHINOOK) it recreated the exact image. This info may help figure out the issue.

**Reply by MaxUint** [reply]
Same, it has completely broken my models and I am getting blank images - something fundemental changed with this update and I have to roll back and will be disabling auto updates in the future

**Reply by cassiebug23** [reply]
I tested multiple models, still noticing a huge change in my photos. I will be rolling back to the older version as soon as I can figure out how. lol

**Reply by MaxUint** [reply]
> I tested multiple models, still noticing a huge change in my photos. I will be rolling back to the older version as soon as I can figure out how. lol

download source code https://github.com/lllyasviel/Fooocus/releases/tag/2.3.1
make a new folder and copy across python_embeded and run.bat, edit  the "entry_with_update.py" script  and remove the try-catch block that encapsulates the update process

**Reply by Krashlog** [reply]
I've posted a new discussion before seeing this (sorry!), but I'm having the same issues. I tried regenerating by taking the metadata straight from the logs, maybe 20 gens or so, and none of them have given me the same results. I'm using the default model (JuggV8), same seed, same prompts, etc. Slightly NSFW but nothing too extreme either, plus I don't think the filtering would change the actual pictures?

I'll probably revert to 2.3.1 in the meantime.

**Reply by cassiebug23** [reply]
> > I tested multiple models, still noticing a huge change in my photos. I will be rolling back to the older version as soon as I can figure out how. lol
> 
> download source code https://github.com/lllyasviel/Fooocus/releases/tag/2.3.1 make a new folder and copy across python_embeded and run.bat, edit the "entry_with_update.py" script and remove the try-catch block that encapsulates the update process

what exactly am I supposed to delete from the entry_with_update.py file? I deleted some of it (I know nothing about Python) and am getting this error:
C:\Users\cassi\Downloads\Fooocus-2.3.1\Fooocus-2.3.1>.\python_embeded\python.exe -s Fooocus\entry_with_update.py
.\python_embeded\python.exe: can't open file 'C:\\Users\\cassi\\Downloads\\Fooocus-2.3.1\\Fooocus-2.3.1\\Fooocus\\entry_with_update.py': [Errno 2] No such file or directory

**Reply by mashb1t** [reply]
will check the issues, give me a few minutes

**Reply by ILikeToasters** [reply]
I don't think specific to model only because I have now not been able to recreate an image on the same model I was able to just a moment ago. Going to keep testing.

**Reply by MaxUint** [reply]
make sure the path from the batch script to entry_wit_update.py is correct

folder structure for me is 
root/
--Fooocus/ (containing 2.3.1 source code, entry_with_update.py should be in this folder)
--python_embeded/
--run.bat

**Reply by cassiebug23** [reply]
> make sure the path from the batch script to entry_wit_update.py is correct
> 
> folder structure for me is root/ Fooocus/ (containing 2.3.1 source code, entry_with_update.py should be in this folder) python_embeded/ run.bat

are you saying run.bat should be INSIDE of the python_embeded folder? Regardless, I tried that too and I'm still getting the same error. Could you tell me what I'm supposed to be deleting from the entry_with_update file?

**Reply by ILikeToasters** [reply]
@mashb1t Okay I might have something. Why might not have noticed in testing. CLIP skip. For a couple images that did not recreate I changed the CLIP skip to 2 and it recreated the image.

**Reply by MaxUint** [reply]
![image](https://github.com/lllyasviel/Fooocus/assets/49106938/8ebe7041-44d2-4268-a32f-a08970e16ceb)

**Reply by MaxUint** [reply]
same fodler layout as it normally is just with the source code for 2.3.1 inside Fooocus just like it normally is :D

**Reply by cassiebug23** [reply]
> remove the try-catch block that encapsulates the update process

this is the part I don't understand. Can you please tell me what I am supposed to remove?

**Reply by ILikeToasters** [reply]
Try adjusting the clip skip as I mentioned above. That might solve issue for moment.

**Reply by cassiebug23** [reply]
> Try adjusting the clip skip as I mentioned above. That might solve issue for moment.

unfortunately I did try that, it did not solve the issue for me.

**Reply by ILikeToasters** [reply]
Yeah I just tested some more and it works for some... But I think that might hint to where the issue might be.

**Reply by cassiebug23** [reply]
> same fodler layout as it normally is just with the source code for 2.3.1 inside Fooocus just like it normally is :D

got it, for some reason the folder order was messed up. thanks

**Reply by mashb1t** [reply]
Reverted clip_skip integration until this has been figured out so you can still use Fooocus in general. Please pull main again / restart Fooocus.
See https://github.com/lllyasviel/Fooocus/pull/3008

Sorry for the inconvenience and thank you for your valuable feedback, will keep you posted.

**Reply by Krashlog** [reply]
> > Try adjusting the clip skip as I mentioned above. That might solve issue for moment.
> 
> unfortunately I did try that, it did not solve the issue for me.

It did get results much closer to the previous version, but the faces are still noticeably worse.

**Reply by ILikeToasters** [reply]
I figured it was CLIP skip since I did not encounter the last week. Also had an image I did generate real quick this morning when I did try the CLIP skip and It did recreate the one that I had set it to 2 but did not recreate the one I did not touch the slider and it just generated as if CLIP skip was on 2 since it was the same result(same seed). hope that makes sense.

**Reply by cassiebug23** [reply]
> > > Try adjusting the clip skip as I mentioned above. That might solve issue for moment.
> > 
> > 
> > unfortunately I did try that, it did not solve the issue for me.
> 
> It did get results much closer to the previous version, but the faces are still noticeably worse.

I agree, it helped very slightly but the faces in my custom checkpoints and loras no longer look as close to the person as they normally do.

**Reply by Krashlog** [reply]
> Sorry for the inconvenience and thank you for your valuable feedback, will keep you posted.

No problems, thanks for the hard work.👍The revert did take care of the problems for most of my gens, so that's a good place to start your investigation.

**Reply by mashb1t** [reply]
fixed in https://github.com/lllyasviel/Fooocus/pull/3011, took a long time to figure out that Fooocus by default uses clip skip 2 and the functionality was correct, but the config & default values weren't. This is also the reason why clip skip 2 yielded the same results as before and why every image without clip skip was different.

**Comment by mzisa** [comment]
Hello, thanks for the update. I am curretnly facing this issue

Traceback (most recent call last):
  File "D:\Fooocus_win64_2-1-831\Fooocus\entry_with_update.py", line 46, in <module>
    from launch import *
  File "D:\Fooocus_win64_2-1-831\Fooocus\launch.py", line 136, in <module>
    config.default_base_model_name, config.checkpoint_downloads = download_models(
  File "D:\Fooocus_win64_2-1-831\Fooocus\launch.py", line 127, in download_models
    load_file_from_url(url=url, model_dir=config.paths_checkpoints[0], file_name=file_name)
  File "D:\Fooocus_win64_2-1-831\Fooocus\modules\model_loader.py", line 27, in load_file_from_url
    download_url_to_file(url, cached_file, progress=progress)
  File "D:\Fooocus_win64_2-1-831\python_embeded\lib\site-packages\torch\hub.py", line 620, in download_url_to_file
    u = urlopen(req)
  File "urllib\request.py", line 216, in urlopen
  File "urllib\request.py", line 525, in open
  File "urllib\request.py", line 634, in http_response
  File "urllib\request.py", line 563, in error
  File "urllib\request.py", line 496, in _call_chain
  File "urllib\request.py", line 643, in http_error_default
urllib.error.HTTPError: HTTP Error 404: Not Found

**Reply by mashb1t** [reply]
@mzisa make sure disable any proxy/VPN before testing and ensure the model URL exists if you have changed the code.
If this doesn't help please open a new issue and fill out all required fields + provide the URL of the model which can't be found.

**Reply by mzisa** [reply]
Thanks for your help. Actually I found it was due to my presets: it was trying to download the model but the download path was set to a file that belongs to my local filesystem. So no worries, not a bug, my bad

**Comment by DavidDragonsage** [comment]
@mashb1t   - I really appreciate all the hard work you put into this, and it also great to see Fooocus back on track again! I also appreciate the work of everyone else involved in this release. Thank you all!  🎂🥳👍

**Comment by badraymen** [comment]
hi, i'm not expert en coding, but how can i come back to the 2.3 version, i have the same issues :(

**Reply by mashb1t** [reply]
@badraymen solved, please restart Fooocus to let it automatically pull the latest changes.

**Reply by badraymen** [reply]
I think there is still a problem with the new version on the "prompt image" part 
yesterday I worked with fooocus and the same function mentioned above, I managed to obtain truly incredible results in a processing time of 30 seconds, with this new version: image generation takes more than 5 minutes to generate an image and the result is really irrelevant, it would be really kind of you to send me version 2.3, because I have work on it and the delivery is for tomorrow, so I only have this night to finish, if please, otherwise I would be in big trouble with the customer

**Reply by xhoxye** [reply]
https://github.com/lllyasviel/Fooocus/releases
Download 2.3.1 and use a startup script that prohibits updates
`.\python_embeded\python.exe -s Fooocus\launch.py %*
pause`

**Reply by badraymen** [reply]
hi, i'm not expert en coding sorry, i dont understand all of this and thats mean, im so so sorry :( :( :(

**Reply by xhoxye** [reply]
https://github.com/lllyasviel/Fooocus/discussions/1405

**Reply by badraymen** [reply]
from this result : 
![image (2) copy](https://github.com/lllyasviel/Fooocus/assets/113091443/f5b2af15-6d8f-45ae-96cc-e32dfa64d293)

to this : 
![2024-05-26_23-54-31_9963](https://github.com/lllyasviel/Fooocus/assets/113091443/582ee902-75ff-4e77-a29b-b2888a82974b)

**Reply by xhoxye** [reply]
But I myself use the 2.4.0 version and don't feel the change in the time consumed
![QQ截图20240527065550](https://github.com/lllyasviel/Fooocus/assets/129571231/4b4b97b7-06c3-4ebd-ac16-cf8d266ab2ad)

**Reply by mashb1t** [reply]
@badraymen make sure to do one of the following:
- restart Fooocus to get the latest version including the clip skip fix
- use the version without fix (2.4.0) and set clip skip to 2 for identical handling as before
- manually switch to 2.3.1

I could not notice any speed decrease, rather a bump instead due to clip skip and/or hyper-sd.
Make sure you don't accidentally have 2 Fooocus instances open and close all irrelevant applications running in the background.

**Reply by badraymen** [reply]
@mashb1t
I am extremely sorry, yes it was my fault I apologize, I launched fooocus and SDWUI at the same time to use "Clip interrogator" the Prompt generator.
Please a suggestion: is there a possibility of adding "Clip interrogator" to "Fooocus", it would be really great, especially since the extension goes wonderfully with SDXL
Thank you for the remarkable effort and the speed of problem solving and really all my respect to all the people who participated in creating this marvel

**Reply by mashb1t** [reply]
@badraymen Fooocus already has an image describe functionality using BLIP, which you can find in the image input > describe tab.
There are no plans of including other image interrogation / description tools but you're welcome to open a feature request.

**Reply by badraymen** [reply]
_@mashb1t_ with all respect for your efforts to guarantee the user the best practical way to better use your magnificent software, I lived the experience with the "describe" tool but the prompts are very limited, on the other hand I would be happy with the fact of added this extension to fooocus if I know how to launch a suggestion or a ticket

**Reply by mashb1t** [reply]
@badraymen => https://github.com/lllyasviel/Fooocus/issues/new/choose

**Comment by MantisOW** [comment]
yesterday I had problems with my faceswaps with inpaint.... and I didn't even have the usual results... is there any way I can change the settings back or the version?

**Reply by mashb1t** [reply]
@MantisOW the issue has been resolved, please restart Fooocus to get the automated update.

**Reply by MantisOW** [reply]
thank you, will try this and hope for the best 👍

**Comment by bobgus39** [comment]
what happens with gradio's live server? it stops working right now.

Thanks.

**Reply by AnalyticsXXI** [reply]
> what happens with gradio's live server? it stops working right now.

Yes, you can check it here https://status.gradio.app/
Usually such things lasted couple of hours.
